### PR TITLE
Feature/98 transaction history

### LINKED ISSUE #98: Get the transaction history for a voter
Get the transaction history of a passport and display it in a understandable fashion.
When no transactionhistory is found display an error that the document is not registered for digital voting.

### PR DESCRIPTION
**Pull request by jonathanraes**
Closes #98 
getTransactions() is still a very long method. Suggestions on how to split it up are welcome.

### CODE REVIEW

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java [hunk 1]
```diff
@@ -75,14 +97,109 @@ public void startDownload() {
...
+        for(Asset a : getAssets()) {
+            if(a.getName().equals(asset.getName())){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Load transactions that involve the given public key, either incomming or outgoing.
```
Comment (wkmeijer) [inline app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: incomming = incoming

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java [hunk 2]
```diff
@@ -75,14 +97,109 @@ public void startDownload() {
...
+        List<Transaction> result = new ArrayList<>();
+        Address address = Address.fromBase58(params, MultiChainAddressGenerator.getPublicAddress(version, Long.toString(addressChecksum), pubKey));
+        String myAddress = address.toString();
+        Set<org.bitcoinj.core.Transaction> ts = kit.wallet().getTransactions(false);
+        for (org.bitcoinj.core.Transaction t:ts) {
+            if (!t.isCoinBase()){
+                boolean checkedInputs = false;
+                for (TransactionOutput o : t.getOutputs()) {
+                    boolean sentToAddr = o.getScriptPubKey().isSentToAddress();
+                    boolean isReturn = o.getScriptPubKey().isOpReturn();
```
Comment (wkmeijer) [inline app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: If I understand correctly isOpReturn gives a boolean on whether or not some extra data is added to the transaction. Why wouldn't you want to display these transactions? (even though I don't think we create any of these)

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java [hunk 3]
```diff
@@ -75,14 +97,109 @@ public void startDownload() {
...
+                                for (TransactionInput in : t.getInputs()) {
+                                    Address fromAddress = in.getScriptSig().getFromAddress(params);
+                                    if (myAddress.equals(fromAddress.toString())) {
+                                        Date date = t.getUpdateTime();
+                                        com.digitalvotingpass.transactionhistory.Transaction newTransaction =
+                                                new com.digitalvotingpass.transactionhistory.Transaction("Sent " + amount, date, "To " + translateAddress(toAddress.toString()));
+                                        result.add(newTransaction);
+                                    }
+                                    checkedInputs = true;
+                                }
```
Comment (wkmeijer) [inline app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: For me it was confusing at first what this if-block does, I think this is a good candidate to move to a separate method.

The if-block below can also be moved to a separate method.

File: app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java [hunk 4]
```diff
@@ -1,41 +1,81 @@
...
+        @Override
+        public void run() {
+            try {
+                Bundle extras = getIntent().getExtras();
+                PublicKey pubKey = (PublicKey) extras.get("pubKey");
+
+                SharedPreferences sharedPrefs = getSharedPreferences(getString(R.string.shared_preferences_file), Context.MODE_PRIVATE);
+                Gson gson = new Gson();
+                String json = sharedPrefs.getString(getString(R.string.shared_preferences_key_election), "");
+                Asset mcAsset = gson.fromJson(json, Election.class).getAsset();
```
Comment (wkmeijer) [inline app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java]: We use these few lines so much, maybe we should write a static function for getting the Election in the Util class. What do you think?

File: app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java [hunk 5]
```diff
@@ -1,41 +1,81 @@
...
+
+                final List<Transaction> transactions = BlockChain.getInstance(null).getTransactions(pubKey, mcAsset);
+                Collections.sort(transactions);
```
Comment (wkmeijer) [inline app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java]: On what does it sort? Maybe it is better to give an argument to force sort in date. (Even though I think it actually does sort on date currently)

**Review by wkmeijer — DISMISSED** [review]
- Like you suggested earlier: rename the Transaction class in the app, so it can be distinguished from Transaction in bitcoinj.
- Wouldn't it be better to implement this method in bitcoinJ, like getAssetBalance() and getAvailableAssets()?
- For my suggestions on refactoring the getTransactions() method: see the comments in the method

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java (hunk 2 above)
Comment (jonathanraes) [reply app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: I think isOpReturn returns whether an output is returning to the sender, and we dont want those

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java (hunk 3 above)
Comment (jonathanraes) [reply app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: Yes I agree its weird, the boolean is there so the input in only checked once per transaction, and not for every output.

File: app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java (hunk 4 above)
Comment (jonathanraes) [reply app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java]: I agree, though maybe a Preferences.getAsset method, in a different class dedicated to preferences

File: app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java (hunk 5 above)
Comment (jonathanraes) [reply app/src/main/java/com/digitalvotingpass/transactionhistory/TransactionHistoryActivity.java]: It does, due to the onCompare implementation in Transaction class that compares the dates

File: app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java (hunk 2 above)
Comment (wkmeijer) [reply app/src/main/java/com/digitalvotingpass/blockchain/BlockChain.java]: http://www.multichain.com/qa/1134/op_return-explanation

File: app/src/main/res/layout/activity_transaction_history.xml [hunk 6]
```diff
@@ -42,7 +42,7 @@
             <!--android:layout_height="wrap_content" />-->
 
         <TextView
-            android:text="No transactions to be displayed"
+            android:text="No transactionHistoryItems to be displayed"
```
Comment (wkmeijer) [inline app/src/main/res/layout/activity_transaction_history.xml]: Please move this hardcoded string to the strings.xml

File: app/src/main/res/layout/activity_transaction_history.xml [hunk 7]
```diff
@@ -15,15 +15,36 @@
...
+    <LinearLayout
+        android:id="@+id/no_transactions"
+        android:visibility="gone"
+        android:orientation="horizontal"
+        android:layout_marginTop="50dp"
+        android:gravity="center"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content">
+        <TextView
+            android:text="No transactionHistoryItems to be displayed"
```
Comment (wkmeijer) [inline app/src/main/res/layout/activity_transaction_history.xml]: Please move this hardcoded string to the strings.xml

**Comment by jonathanraes** [comment]
https://github.com/digital-voting-pass/digital-voting-pass-bitcoinj/commit/49e20e1b52023ebcef9e10c3f859e7ec78f53aef

File: app/src/main/java/com/digitalvotingpass/digitalvotingpass/ResultActivity.java [hunk 8]
```diff
@@ -244,7 +247,7 @@ public void nextVoter() {
      * TODO: implement this method
      */
     public void confirmVote() {
-        Toast.makeText(this, "Transaction sent", Toast.LENGTH_LONG).show();
+        Toast.makeText(this, "TransactionHistoryItem sent", Toast.LENGTH_LONG).show();
```
Comment (klikooo) [inline app/src/main/java/com/digitalvotingpass/digitalvotingpass/ResultActivity.java]: TransactionHistoryItem sent?

**Comment by jonathanraes** [comment]
continued in #126

**Pull request closed without merging**